### PR TITLE
Add Deep-Learning-using-PyTorch.md

### PR DESCRIPTION
**Pull request by yunkon-kim**
1주일 동안 수강했던 "PyTorch를 활용한 Deep Learning" 강좌 중 일부분을 정리하였습니다. (복기하고 공유하는 차원으로 작성한 글입니다.)

### CODE REVIEW

**Comment by yunkon-kim** [comment]
@white255, @akakakakakaa 안녕하세요 ^^

리뷰를 부탁드립니다 ~~

**Comment by seokho-son** [comment]
거의.. 책을 쓰셨군요.. ㅎㅎ 👍

File: docs/Deep-Learning-using-PyTorch.md [hunk 1]
```diff
@@ -0,0 +1,760 @@
...
+```python
+# 위 실습의 설명 참고
+def cost(x,y,w):
+    c=0
+    for i in range (len(x)):
+        hx = w*x[i]
+        c = c + (hx-y[i])**2
+    return c/len(x)
+
+# 기울기를 구하는 함수
```
Comment (white255) [inline docs/Deep-Learning-using-PyTorch.md:165]: 기울기를 사용한 w 평균 변화량 구하는 함수

File: docs/Deep-Learning-using-PyTorch.md [hunk 2]
```diff
@@ -0,0 +1,760 @@
...
+    c = cost(x_data, y_data, w)
+    print('epoch:', epoch, 'cost=', c, 'w=', w)
+    g = gradient(x_data, y_data, w)
+    w = w - 0.1 * g
+print('최종 w: ', w)
+
+```
+
+합성함수 미분을 통해 Loss function `(hx-y[i])**2`의 기울기를 구하면 (`hx = w*x[i]`),
+`(hx-y[i])*x[i]`을 도출할 수 있습니다. (상수는 영향력이 미비하여 날린 것일까요? :thinking:)
```
Comment (white255) [inline docs/Deep-Learning-using-PyTorch.md:186]: (상수는 영향력이 미비하여 날린 것일까요? :thinking:) -> 형이 정의한 식에는 상수가 없습니다. 그래서 상수에 대한 식이 발생하지 않습니다

File: docs/Deep-Learning-using-PyTorch.md [hunk 3]
```diff
@@ -0,0 +1,760 @@
...
+b = torch.tensor(10.0, requires_grad=True)
+
+# 
+optimizer = o.Adam([w,b], lr=0.1)
+for i in range(2000):
+    hx = w*x+b                    # Forward 개념
+    cost = torch.mean((hx-y)**2)  # cost 개념 
+    optimizer.zero_grad()         # 기존 미분 값 초기화 (미분 값이 합산되는 것을 방지)
+    cost.backward()               # 미분 w, b (왜 미분인지 추후 설명)
+    optimizer.step()              # w = w - 0.1 * 미분 값, b = b - 0.1 * 미분 값
```
Comment (white255) [inline docs/Deep-Learning-using-PyTorch.md:477]: w = w - 0.1 * 미분 값, b = b - 0.1 * 미분 값
-> w, b를 cost를 적용한 업데이트
-> optimizer를 adam을 썼기 때문에 위 식이 적용 되지 않습니다

**Review by white255 — COMMENTED** [review]
리뷰 의견 남깁니다.

용어와 관련된 부분에서만 조금 수정 하시면 될 것 같습니다.

소스는 문제 없습니다.

File: docs/Deep-Learning-using-PyTorch.md (hunk 1 above)
Comment (yunkon-kim) [reply docs/Deep-Learning-using-PyTorch.md:165]: 리뷰 고맙습니다 ^^
역시 날카롭게 봐주셨네요! "기울기", "w를 구하는 것"을 전체적으로 구분없이 혼용했었네요. 수정 반영 하겠습니다 ^^

File: docs/Deep-Learning-using-PyTorch.md (hunk 2 above)
Comment (yunkon-kim) [reply docs/Deep-Learning-using-PyTorch.md:186]: @white255 

상수로 잘못 서술 했습니다 ^^;; 말이 나온김에 질문을 드려야겠네요 ㅎㅎ

모호했던 부분은요. 합성함수 미분을 했으니 제곱수 2가 내려와서 계수로 곱해질 것 같은데요. 이것을 유지하는 예제도 있고, 설명 없이 2로 나누어 없애는 경우도 있어서 궁금하기도 했던 부분 입니다.

대체로 이렇게 하는 방식까요? 아니면 개발자가 알아서 튜닝하는 부분 일까요?

File: docs/Deep-Learning-using-PyTorch.md (hunk 3 above)
Comment (yunkon-kim) [reply docs/Deep-Learning-using-PyTorch.md:477]: 주석을 수정하겠습니다 ~~

**Comment by yunkon-kim** [comment]
@white255 

가볍게 리뷰를 부탁드렸는데 엄청난 진심을 담아 리뷰해주셨네요. 정말 감사드립니다 ❤️

File: docs/Deep-Learning-using-PyTorch.md [hunk 4]
```diff
@@ -0,0 +1,760 @@
...
+### Math, engineering approach, and machine learning
+
+Machine learning (ML)을 처음 접했을 때도, 또한 이번 강좌에서도 가장 먼저 만나게 된 것은 아래 수식이었습니다. ( $w$ : Weight, $b$ : Bias )
+
+$$h(x) = w x + b$$
+
+Hypothesis(가설)이라 불리는 이 수식을 방정식 또는 데이터를 토대로 도출한 것 정도로 지극히 추상적으로 바라봤는데요.
+강좌를 들은 후 다른 관점으로 보게 되었습니다. 
+전에는 몇 개의 $w$, $b$ 쌍을 주고 $x$, $y$를 찾는 것으로 생각했던 것을 
+지금은 데이터라는 많은 수의 $x$, $y$ 쌍을 줬을 때 $w$, $b$를 찾는 것이라고 이해할 수 있습니다.
```
Comment (seokho-son) [inline docs/Deep-Learning-using-PyTorch.md:28]: @yunkon-kim 
한번 읽어보고 있습니다. ㅎㅎ
해당 부분은 개인적인 경험이 들어가 있어서, 처음 보시는 분들께는 오히려 혼동을 일으킬 수 있지 않을까.. 하는 생각도 들었습니다.
특히 "몇 개의 $w$, $b$ 쌍을 주고 $x$, $y$를 찾는 것으로 생각했던 것을" 요부분.. :)

조금 일반적인 설명으로 개선해보면 어떨까요?

바로 아래 설명이 조금 더 명확한 느낌입니다. :)

우리는 주어진 세 점을 분석하여 $y = x + 1$라는 수식을 찾아내고, 이를 프로그래밍에 활용했습니다.
지금은 세 점 (1, 2), (2, 3), (3, 4)과 가설 $y = w x + b$을 입력해주고 컴퓨터 프로그램이 학습을 통해 만족하는 $w$와 $b$를 찾아내고, 이를 프로그램에 활용하는 방식으로 달라진 것이랄까요?

File: docs/Deep-Learning-using-PyTorch.md [hunk 5]
```diff
@@ -0,0 +1,760 @@
...
+
+예를 들어, 기존에는 세 점 (1, 2), (2, 3), (3, 4)가 주어졌을 때, 우리는 주어진 세 점을 분석하여 $y = x + 1$라는 수식을 찾아내고, 이를 프로그래밍에 활용했습니다.
+지금은 세 점 (1, 2), (2, 3), (3, 4)과 가설 $y = w x + b$을 입력해주고 컴퓨터 프로그램이 학습을 통해 만족하는 $w$와 $b$를 찾아내고, 이를 프로그램에 활용하는 방식으로 달라진 것이랄까요?
+
+한편으로는 "새로운 문제 풀이 방법론, 역할, 생태계가 추가되었구나!"라는 생각마저 들었습니다 ^^;; 
+기존에는 수학자가 현상/데이터를 분석하여 적절한 수식을 도출하거나, 개발자가 엔지니어링 접근법을 통해 적절한 수식 $x$를 찾고, 이를 프로그래밍에서 사용했다면,
```
Comment (seokho-son) [inline docs/Deep-Learning-using-PyTorch.md:34]: 새로운 문제 풀이 방법론이 나왔다고 하기에는
다소 역사가 긴 방법이라, 조금 더 완곡하게 표현하면 좋지 않을까 합니다. :)
(수치최적화, 머신 러닝 등.. 오래된 학문의 기법들이라)
구/신 방식이 아닌, 원래 2가지 방법론이 있는 것 정도로?!

File: docs/Deep-Learning-using-PyTorch.md [hunk 6]
```diff
@@ -0,0 +1,760 @@
...
+
+<p align="center">
+  <img src="https://user-images.githubusercontent.com/7975459/195769402-04014982-180d-49c1-af1d-98c9f9459be9.png" width="50%" height="50%" />
+</p>
+
+### 상관관계 분석
+
+두 변수 간에 어떤 선형적 관계를 가졌는지를 분석하는 방법이다. 
+피어슨 상관계수 (Pearson correlation coefficient 또는 Pearson's r)는 두 변수 X와 Y 간의 선형 상관 관계를 계량화한 수치를 나타낸다. 
+r 값은 X 와 Y가 완전히 동일하면 +1, 전혀 다르면 0, 반대 방향으로 완전히 동일하면 –1 을 가진다.
```
Comment (seokho-son) [inline docs/Deep-Learning-using-PyTorch.md:504]:
```suggestion
두 변수 간에 어떤 선형적 관계를 가졌는지를 분석하는 방법입니다. 
피어슨 상관계수 (Pearson correlation coefficient 또는 Pearson's r)는 두 변수 X와 Y 간의 선형 상관 관계를 계량화한 수치를 나타냅니다. 
r 값은 X 와 Y가 완전히 동일하면 +1, 전혀 다르면 0, 반대 방향으로 완전히 동일하면 –1 을 가집니다.
```

**Review by seokho-son — COMMENTED** [review]
:)

File: docs/Deep-Learning-using-PyTorch.md (hunk 5 above)
Comment (yunkon-kim) [reply docs/Deep-Learning-using-PyTorch.md:34]: 동감합니다 ㅎㅎ 완곡하게 수정하겠습니다~~ ^^

File: docs/Deep-Learning-using-PyTorch.md (hunk 4 above)
Comment (seokho-son) [reply docs/Deep-Learning-using-PyTorch.md:28]: @yunkon-kim 특별히 리뷰를 하는 것은 아니니~ 원작자께서 자유롭게.. :)

File: docs/Deep-Learning-using-PyTorch.md (hunk 2 above)
Comment (yunkon-kim) [reply docs/Deep-Learning-using-PyTorch.md:186]: @white255 감사합니다~~

Step 횟수, Learning rate 값이 더 많은 영향을 주겠네요 ^^

**Review by white255 — APPROVED** [review]
확인하였습니다

**Comment by yunkon-kim** [comment]
@white255 감사합니다 ^^
